### ISSUE TITLE
Intercepting routes not working with QueryString (pagination)

### ISSUE THREAD

**Issue by zaru** [vercel/next.js#56575]
### Link to the code that reproduces this issue

https://github.com/zaru/nextgram

### To Reproduce

1. Launch the code from https://github.com/zaru/nextgram.
2. Open `/feed` .
3. Click on the link for `<Link> 2` to navigate to the page.
4. Click on any Photo.
5. Instead of being intercepted and displaying a modal, it results in a regular page transition.

### Current vs. Expected behavior

Bug Description:
When using the `<Link>` component to navigate to a page with a QueryString (for pagination, etc.), the Intercepting Routes do not function properly after the transition. However, if you directly open a page with a QueryString and then click on a Photo, it works as expected. Also, when navigating in the same manner using `<a>` instead of `<Link>`, it functions correctly.

Expected Behavior:
I would like the Intercepting Routes to function even when the QueryString is modified using `<Link>` or `router.push()` .

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000
Binaries:
  Node: 18.12.1
  npm: 9.6.6
  Yarn: 3.6.3
  pnpm: N/A
Relevant Packages:
  next: 13.5.7-canary.13
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Routing (next/router, next/navigation, next/link)

**Comment by zaru** [comment]
### Differences from the 'nextgram' in the example:
- Implemented pagination support.
- Changed the URL for the Photo list from / to /feed.

### Description of the Reproduction Video:

https://github.com/vercel/next.js/assets/235650/3463a83f-cb5f-41db-9087-40cedee188aa

#### unexpected behavior
1. Click on Link 2 from /feed.
2. Click on a Photo.
3. Instead of a modal, the regular Photo page is displayed.

#### expected behavior
1. Directly open /feed?page=2.
2. Click on a Photo.
3. The Photo is displayed in a modal.

**Comment by zaru** [comment]
I'm using middleware as a workaround for this issue. It's working fine for now. 
However, I'm not sure if this is the right approach...

```typescript
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const requestHeaders = new Headers(request.headers);
  const nextUrl = requestHeaders.get('Next-Url');
  if (nextUrl) {
    const [path, _] = nextUrl.split('?');
    requestHeaders.set('Next-Url', path);
  }

  return NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  })
}

export const config = {
  matcher: '/:path*',
}
```

**Comment by zaru** [comment]
I tried again with version 14.0.4-canary.6, but it seems that the issue is still not resolved.

**Comment by zaru** [comment]
I have confirmed that this has been fixed in 14.0.4.
